pkg/client/events/replay_client_integration_test.go

```diff
@@ -0,0 +1,150 @@
+//go:build integration
+
+package events_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"cosmossdk.io/depinject"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pokt-network/poktroll/pkg/client/events"
+	"github.com/pokt-network/poktroll/pkg/observable"
+	"github.com/pokt-network/poktroll/testutil/testclient/testeventsquery"
+)
+
+// Create the generic event type and decoder for the replay client
+
+var _ messageEvent = (*tEvent)(nil)
+
+type messageEvent interface {
+	EventMessage() string
+}
+
+type tEvent struct {
+	Message string `json:"message"`
+}
+
+type messageEventReplayObs observable.ReplayObservable[messageEvent]
+
+func (t *tEvent) EventMessage() string {
+	return t.Message
+}
+
+func newDecodeEventMessageFn() events.NewEventsFn[messageEvent] {
+	return func(eventBz []byte) (messageEvent, error) {
+		t := new(tEvent)
+		if err := json.Unmarshal(eventBz, t); err != nil {
+			return nil, err
+		}
+		if t.Message == "" {
+			return nil, events.ErrEventsUnmarshalEvent
+		}
+		return t, nil
+	}
+}
+
+// newMessageEventBz returns a new message event in JSON format
+func newMessageEventBz(eventNum int32) []byte {
+	return []byte(fmt.Sprintf(`{"message":"message_%d"}`, eventNum))
+}
+
+func TestReplayClient_Remapping(t *testing.T) {
+	var (
+		ctx              = context.Background()
+		connClosed       atomic.Bool
+		delayEvent       atomic.Bool
+		readEventCounter atomic.Int32
+		eventsReceived   atomic.Int32
+		eventsToRecv     = int32(10)
+		errCh            = make(chan error, 1)
+		timeoutAfter     = 3 * time.Second // 1 second delay on retry.OnError
+	)
+
+	// Setup the mock connection and dialer
+	connMock, dialerMock := testeventsquery.NewTwiceMockConnAndDialer(t, &connClosed, &delayEvent)
+	// Mock the connection receiving events
+	connMock.EXPECT().Receive().
+		DoAndReturn(func() (any, error) {
+			// Simulate ErrConnClosed if connection is isClosed.
+			if connClosed.Load() {
+				return nil, events.ErrEventsConnClosed
+			}
+
+			// Delay the event if needed
+			if !delayEvent.Load() {
+				time.Sleep(50 * time.Millisecond)
+				delayEvent.CompareAndSwap(false, true)
+			}
+
+			eventNum := readEventCounter.Add(1) - 1
+			event := newMessageEventBz(eventNum)
+			if eventNum == 2 {
+				// Simulate the connection closing
+				connMock.Close()
+			}
+
+			// Simulate IO delay between sequential events.
+			time.Sleep(50 * time.Microsecond)
```

Comment:
I'm worried about different machines having different performance characteristics, so can you explain where/how 50ms was validated?

Reply:
It **needs to be non-zero** because `#Recieve()` is called in tightest loop possible (i.e. max speed limited by go runtime). If this delay is too small, in this environment where there is no real delay between events being available for publishing, calls to `#Receive()` will bog down the runtime, leading to other unexpected errors.

We should **prefer to keep it as small as possible** (individual orders of magnitude matter here) in order to minimize the feedback loop in the event that we need to debug flaky behavior in the future; I typically consider a test likely to be flake-free if I can get it to pass 10K times contiguously.

Reply:
IIRC, there are a few other places in tests where we have to slow down some loops for similar reasons.

Reply:
Yeah there needs to be a delay but its not super important on the timing of the delay the only delay that has some significance is the delayed event (line 81) as this gives the events query client time to subscribe

Reply:
#PUC with bryan's explanation